### PR TITLE
Add `apply` to `asynctools`

### PR DESCRIPTION
**Pull request by mristin**
* Add `apply` to `asynctools` module

* Document `apply`

* Fixes the issue #33.

### CODE REVIEW

**Comment by mristin** [comment]
@maxfischer2781 I need to rebase this pull request after we merge in the pull request #33, but you can already give it a first pass and review it.

File: docs/source/api/asynctools.rst [hunk 1]
```diff
@@ -19,3 +19,6 @@ Iterator lifetime
 
 .. autofunction:: scoped_iter(iterable: (async) iter T)
     :async-with: :async iter T
+
+.. autofunction:: apply(func: Callable[..., T], *args: Awaitable[Any], **kwargs: Awaitable[Any]) -> T
+    :async:
```
Comment (maxfischer2781) [inline docs/source/api/asynctools.rst:24]:
```suggestion
.. autofunction:: apply(func: (*T, **T) -> R, *args: await T, **kwargs: await T) -> R
    :async:
```

File: asyncstdlib/asynctools.py [hunk 2]
```diff
@@ -174,3 +177,33 @@ async def head_tail(iterable, leading=5, trailing=5):
     if not hasattr(iterator, "aclose"):
         return nullcontext(iterator)
     return _ScopedAsyncIteratorContext(iterator)
+
+
+async def apply(
+    func: Callable[..., T], *args: Awaitable[Any], **kwargs: Awaitable[Any]
```
Comment (maxfischer2781) [inline asyncstdlib/asynctools.py:183]: Allow using ``func`` as a keyword.
```suggestion
    __func: Callable[..., T], *args: Awaitable[Any], **kwargs: Awaitable[Any]
```

File: asyncstdlib/asynctools.py [hunk 3]
```diff
@@ -174,3 +177,33 @@ async def head_tail(iterable, leading=5, trailing=5):
...
+        result = await apply(
+            lambda x, y: x ** y,
+            compute_something(),
+            compute_something_else())
+
+    The function ``apply`` serves, for example, a practical use case when you want to
+    chain operations on awaitables and need to pass around the final awaitable
+    for further operations.
+    """
+    return func(
```
Comment (maxfischer2781) [inline asyncstdlib/asynctools.py:207]:
```suggestion
    return __func(
```

**Review by maxfischer2781 — CHANGES_REQUESTED** [review]
I recommend changing the function argument name to avoid collisions.

Do you feel a fast-path is required? Say, if ``kwargs`` are empty?

**Comment by mristin** [comment]
> Do you feel a fast-path is required? Say, if `kwargs` are empty?

I am not familiar with the concept of *fast-path* -- you mean checking if `kwargs` are empty and branching accordingly?

**Comment by mristin** [comment]
@maxfischer2781 The pull request is ready for review (and merge, if all's good).

**Review by maxfischer2781 — APPROVED** [review]
Looks good to me. Many thanks for the contribution.

**Pull request merged**